src/stan/math/prim/mat/meta/index_type.hpp

```diff
@@ -21,6 +21,16 @@ namespace stan {
       typedef typename Eigen::Matrix<T,R,C>::Index type;
     };
 
+    template <typename T>
+    struct index_type<Eigen::MatrixBase<T> > {
+      typedef typename index_type<T>::type type;
+    };
+
+    template <typename T>
+    struct index_type<Eigen::Map<T> > {
```

Comment:
> On Mar 12, 2015, at 11:25 AM, Daniel Lee <notifications@github.com> wrote:
> 
> ...
> Sorry, I get confused with Eigen doc. I still haven't found how to specialize matrix multiplication. The question I have is... is there a single templated function we can write to cover both Eigen::MatrixBase and Eigen::Map?

The page to read is this:

  http://eigen.tuxfamily.org/dox/TopicFunctionTakingEigenTypes.html

MatrixBase includes Map.  DenseBase includes both MatrixBase and ArrayBase.

The Eigen doc itself isn't very good at saying what class(es)
a given class inherits from.

But now I'm really confused, because what they're saying should work
doesn't, unless I'm missing something fundamental.  Could this be a bug
in their asserts?


#include <gtest/gtest.h>
#include <iostream>
#include <Eigen/Dense>

using Eigen::Map;
using Eigen::Matrix;
using Eigen::MatrixXd;
using Eigen::MatrixBase;
using Eigen::Dynamic;

template <typename Derived>
double foo(const MatrixBase<Derived>& a, int m, int n) {
  return a(m,n);
}

TEST(foo, bar) {

  MatrixXd z(2,3);
  z << 1, 2, 3,
    4, 5, 6;
  EXPECT_FLOAT_EQ(1, foo(z,0,0)); // OK
  EXPECT_FLOAT_EQ(6, foo(z,1,2)); // OK

  std::vector<double> x;
  for (int i = 0; i < 6; ++i)
    x.push_back(i);
  Map<MatrixXd> y(&x[0],2,3);
  EXPECT_FLOAT_EQ(0, foo(y,0,0));  // OK
  EXPECT_FLOAT_EQ(5, foo(y,1,2));  // OK

  MatrixXd v(2,3);
  v << 1, 2, 3, 4, 5, 6;

  EXPECT_FLOAT_EQ(6, foo(v.transpose(),2,1)); // OK

  MatrixXd u(2,2);
  u << 1, 2, 3, 4;

  MatrixXd uv = u * v;
  EXPECT_FLOAT_EQ(33, foo(uv,1,2));  // OK

  std::cout << "(u * v)=" << std::endl
            << (u * v) << std::endl;

  std::cout << "(u * v).rows()=" << ((u * v).rows()) << std::endl;
  std::cout << "(u * v).cols()=" << ((u * v).cols()) << std::endl;

  std::cout << std::endl;

  EXPECT_FLOAT_EQ(9, foo((u * v).transpose(), 0, 0)); // OK

  EXPECT_FLOAT_EQ(9, foo((u * v), 0, 0)); //  FAILS.  WTF?
}


I'm going to write to the Eigen list and see if they have any suggestions.

- Bob

> …
> On Wed, Mar 11, 2015 at 8:09 PM, Bob Carpenter <notifications@github.com> wrote: In src/stan/math/prim/mat/meta/index_type.hpp <https://github.com/stan-dev/stan/pull/1210#discussion_r26268589>: > @@ -21,6 +21,16 @@ namespace stan { > typedef typename Eigen::Matrix<T,R,C>::Index type; > }; > > + template <typename T> > + struct index_type<Eigen::MatrixBase<T> > { > + typedef typename index_type<T>::type type; > + }; > + > + template <typename T> > + struct index_type<Eigen::Map<T> > { Yes. I'd suggest looking these things up! Eigen has great doc. Map<T> is the expression template class that maps contiguous memory to an Eigen type T. - Bob … <#14c0b507620a87f6_> On Mar 12, 2015, at 7:24 AM, Daniel Lee <notifications@github.com> wrote: In src/stan/math/prim/mat/meta/index_type.hpp: > @@ -21,6 +21,16 @@ namespace stan { > typedef typename Eigen::Matrix<T,R,C>::Index type; > }; > > + template <typename T> > + struct index_type<Eigen::MatrixBase<T> > { > + typedef typename index_type<T>::type type; > + }; > + > + template <typename T> > + struct index_type<Eigen::Map<T> > { Is Eigen::Map<T> different than Eigen::MatrixBase<T>? — Reply to this email directly or view it on GitHub. — Reply to this email directly or view it on GitHub <https://github.com/stan-dev/stan/pull/1210/files#r26268589>.
> —
> Reply to this email directly or view it on GitHub.
>